arangod/Aql/Optimizer/Rule/OptimizerRuleVectorIndex.cpp

```diff
@@ -0,0 +1,303 @@
+      approxFunctionParameters->getMemberUnchecked(1);
+
+  if (approxFunctionParamLeft->type ==
+      arangodb::aql::NODE_TYPE_ATTRIBUTE_ACCESS) {
+    if (checkApproxNearVariableInput(vectorIndex, approxFunctionParamLeft,
+                                     outVariable)) {
+      return approxFunctionParamRight;
+    }
+  }
+  if (checkApproxNearVariableInput(vectorIndex, approxFunctionParamRight,
+                                   outVariable)) {
+    return approxFunctionParamLeft;
+  }
+
+  return nullptr;
+}
+
+}  // namespace
+
+std::vector<std::shared_ptr<Index>> getVectorIndexes(
+    auto& enumerateCollectionNode) {
+  std::vector<std::shared_ptr<Index>> vectorIndexes;
+  auto indexes = enumerateCollectionNode->collection()->indexes();
+  std::ranges::copy_if(
+      indexes, std::back_inserter(vectorIndexes), [](auto const& elem) {
+        return elem->type() == Index::IndexType::TRI_IDX_TYPE_VECTOR_INDEX;
+      });
+
+  return vectorIndexes;
+}
+
+void arangodb::aql::useVectorIndexRule(Optimizer* opt,
+                                       std::unique_ptr<ExecutionPlan> plan,
+                                       OptimizerRule const& rule) {
+  bool modified{false};
+
+  containers::SmallVector<ExecutionNode*, 8> nodes;
+  // avoid subqueries for now
+  plan->findNodesOfType(nodes, EN::ENUMERATE_COLLECTION, true);
+  for (ExecutionNode* node : nodes) {
+    auto* enumerateCollectionNode =
+        ExecutionNode::castTo<EnumerateCollectionNode*>(node);
+
+    // check if there are vector indexes on collection
+    auto const vectorIndexes = getVectorIndexes(enumerateCollectionNode);
+    if (vectorIndexes.empty()) {
+      continue;
+    }
+
+    // check that enumerateColNode has both sort and limit
+    auto* currentNode = enumerateCollectionNode->getFirstParent();
+    // skip over some calculation nodes
+    while (currentNode != nullptr &&
+           currentNode->getType() == EN::CALCULATION) {
+      currentNode = currentNode->getFirstParent();
+    }
+
+    if (currentNode == nullptr || currentNode->getType() != EN::SORT) {
+      LOG_RULE << "DID NOT FIND SORT NODE, but instead "
+               << (currentNode ? currentNode->getTypeString() : "nullptr");
+      continue;
+    }
+    auto* sortNode = ExecutionNode::castTo<SortNode*>(currentNode);
+
+    auto const* maybeLimitNode = sortNode->getFirstParent();
+    if (!maybeLimitNode || maybeLimitNode->getType() != EN::LIMIT) {
+      LOG_RULE << "DID NOT FIND LIMIT NODE, but instead "
+               << (maybeLimitNode ? maybeLimitNode->getTypeString()
+                                  : "nullptr");
+      continue;
+    }
+
+    // check LIMIT NODE
+    auto const* limitNode =
+        ExecutionNode::castTo<LimitNode const*>(maybeLimitNode);
+    // Offset cannot be handled, and there must be a limit which means topK
+    if (limitNode->offset() != 0 || limitNode->limit() == 0) {
+      LOG_RULE << "Limit not set";
+      continue;
+    }
```

Comment:
I think we can, this was jut left unimplemented. I will do it now. I don't think it should be much of a problem

Reply:
I don't think we should implement this. This can only cause trouble in the long run.
1. `APPROX_NEAR` does not provide any stable sort. Pagination is not going to work. ArangoDB has no stable result set unless the sorting is strictly unique.
2. This is not well supported by faiss. For `LIMIT n, k` we always have to compute the top n +k and drop the first n.
3. No other vector database supports this, the usecase is limited.

Reply:
1. It is stable per query parameters and the vector index definition, the repeated query will produce identical results
2. I agree, but I don't see other way of doing this
3. I think one of the simplest cases would be setting off first result to not get the identical vector back, e.g. LIMIT 1, 5

But I think we would need to circumvent the Limit node and handle that in our own EnumerateVectorIndexExecutor. Since we handle the limit there as well.

Reply:
Regarding 1. it is not in the cluster because even if the merge is sorted, there is no deterministic tie breaker for equal distant vectors.

Reply:
Lars and Tobias have talked about this and agreed that we should support this for the sake of AQL semantics and to stay consistent